Prep for Docker log drivers announcement

tgross opened the pull request:
@misterbisson I'm opening this PR for review, but I'll want to add some more content for the README before merging. This will complement what'll probably be more extensive work in `triton-elk` that I'll be working on through early next week. The current master branch prior to merging this was tagged as `preview`.

Changes:
- switch to Alpine Linux
- install ES from the source tarball because there's no APK, which required a lot of fiddly Dockerfile
- check package signatures for ES and Containerbuddy during build
- discover Consul via Triton CNS when deployed to Triton (but works locally via envvar)
- consolidate the Docker Compose files to take advantage of getting rid of `--link`s in the main file

tgross committed:
Consolidate compose files

Drop common-compose.yml by moving its services into docker-compose.yml.
Update local-compose.yml to extend the main (on-Triton) docker-compose.yml
file instead.

tgross committed:
Check signatures of binaries in Dockerfile

tgross committed:
Discover Consul via Triton CNS

Use an environment variable to pass the Triton CNS svc location to
Containerbuddy. By also adding the service name via ES_SERVICE_NAME environment
variable we can also eliminate the second Containerbuddy config file and all our
`--link`s on Triton.

tgross committed:
Switch to Alpine Linux base

tgross committed:
Clarify the role of the makefile for local dev only

tgross commented:
Another note is that I considered trying to adapt the primary-replica bootstrap that we did for `triton-mysql` but there are a couple complications to that:

- ES doesn't have a simple primary-replica setup; there can be multiple masters, data-only nodes, and nodes that serve as both masters and data. How you divide that up is highly idiosyncratic to the workload you have.
- given how long `triton-mysql` took, it's probably more work than we probably want to invest in this blueprint right now. :grinning:

misterbisson commented:
> given how long `triton-mysql` took

Agreed.

> ES doesn't have a simple primary-replica setup

Well, this is relatively easy if we can eliminate the hybrid master-data nodes, no?

```yml
es-master:
    image: elasticsearch
    command: do-master

es-data:
    image: elasticsearch
    command: do-data
```

They're both the same image, but a different command to start. 

On starting an existing master/cluster, it will look for an existing master, and if it finds one, join it as a peer. This is similar to our Couchbase work. If there is no existing cluster, it will form one. So, the masters are an upstream for each other.

On starting a data node, they also look for an existing cluster, but fail if there is none. They, too, have the masters as upstreams.

OTOH, this also easy if we break that single Compose file into three. They can still have the same behaviors as described above (hybrid nodes would mostly behave like masters, I guess), but you wouldn't be forced to start one of each at first. The big difference, I guess, is that hybrid nodes would announce both the master and data services, no?

**I'm probably wrong about all of that, however, and I stand by the point at the top: out of scope.**

tgross committed:
Bump Containerbuddy to 1.2.0

tgross committed:
Make sure CNS tag is DNS-friendly

misterbisson commented on Dockerfile:
```diff
@@ -1,42 +1,51 @@
-FROM debian:jessie
+FROM alpine:3.3
```
Depends on `OS-5184`, as discussed elsewhere. We should ticket a separate issue to update this when `OS-5184` is shipped.

tgross committed:
Revert to Debian from Alpine, close off 9200

tgross committed:
Fix clustering of master+data nodes

- Assign master+data nodes to elasticsearch-master service name
- Change onStart behavior to allow master+data nodes to retry a few times
  if there is no other master yet.

tgross commented:
@misterbisson this and https://github.com/tgross/triton-elk/pull/2 are probably ready for re-review at this point.

misterbisson commented on Dockerfile:
```diff
@@ -10,21 +10,23 @@ RUN apt-get update && \
+    && rm /tmp/containerbuddy.tar.gz
+
+# Create and take ownership over required directories
 # Take ownership over required directories
```
Dupe comment?

tgross replied:
Copy-pasta bug! Fixed!

misterbisson commented:
Solid. Small comment, but :walking: :house_with_garden:.

tgross committed:
Remove redundant comment
